benchmarks/distributed/ddp/compare/python_ddp.py

```diff
@@ -0,0 +1,228 @@
+    LegacyDistributedDataParallel.
+
+    Same as DistributedDataParallel to use this class , a process group needs to
+    be initiated.
+
+    Example::
+
+        >>> torch.distributed.init_process_group(
+        >>>     backend='gloo', world_size=N, init_method='...'
+        >>> )
+        >>> pg = dist.distributed_c10d._get_default_group()
+        >>> async_reduction = True
+        >>> module = ToyModel()
+        >>> ddp_model = PythonDDP(module, pg, async_reduction)
+        >>> loss_fn = nn.MSELoss()
+        >>> optimizer = optim.SGD(ddp_model.parameters(), lr=0.001)
+        >>> outputs = ddp_model(torch.randn(20, 10).to(rank))
+        >>> labels = torch.randn(20, 10).to(rank)
+        >>> loss_fn(outputs, labels).backward()
+        >>>
+        >>> # Reduce param grads
+        >>> ddp_model.all_reduce_grads()
+        >>> optimizer.step()
+        >>>
+
+    """
+
+    class Bucket:
+        """Bucket is a container for list of params. """
+
+        def __init__(self, max_buffer_size):
+            self.param_to_offset = {}
+            self.buffer = None
+            self.ready_param_grad_count = 0
+            self.total_elements = 0
+            self._MAX_BUFFER_SIZE = max_buffer_size
+
+        def __str__(self):
+            return "Bucket: num_params={}, total_elements={}, ready_param_grad_count={}".format(
+                len(self.param_to_offset),
+                self.total_elements,
+                self.ready_param_grad_count)
+
+        def is_full(self):
+            """
+            Returns whether grad for all the params in current bucket are ready
+            and copied to self.buffer.
+            """
+            assert self.ready_param_grad_count >= 0
+            assert self.ready_param_grad_count <= len(self.param_to_offset)
+            return len(self.param_to_offset) == self.ready_param_grad_count
+
+        def empty(self):
+            self.ready_param_grad_count = 0
+
+        def try_hold_param(self, param):
+            """
+            Checks whether current bucket has enough buffer to hold the incoming
+            param. If there is enough space, distribute param into current
+            bucket and Returns true. Otherwise, returns False.
+            """
+            if self.total_elements + param.numel() <= self._MAX_BUFFER_SIZE :
+                self.param_to_offset[param] = self.total_elements
+                self.total_elements += param.numel()
+                return True
+            else:
+                return False
+
+    def __init__(self, module, process_group, async_reduction=True, buffer_size=2 ** 22):
+        super(PythonDDP, self).__init__()
+
+        self.module = module
+        self.process_group = process_group
+        self.world_size = dist.get_world_size(group=self.process_group)
+        self.async_reduction = async_reduction
+
+        # Holds all_reduce handles, used when async_reduction is True
+        self.async_handles = set()
```

Comment:
Can we try using list instead of a set here? When the bucket size is large, the insertion and scanning of a hash set can be slower than a flat vector.

Instead, use `[None] * bucket_size` should be equivalent to a vector + reserve the size first.